packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,321 @@
+  error TooManyResults();
+
+  /// @notice Create a new Tally contract
+  /// @param verifierContract The Verifier contract
+  /// @param vkRegistryContract The VkRegistry contract
+  /// @param pollContract The Poll contract
+  /// @param mpContract The MessageProcessor contract
+  /// @param tallyOwner The owner of the Tally contract
+  /// @param pollMode The mode of the poll
+  constructor(
+    address verifierContract,
+    address vkRegistryContract,
+    address pollContract,
+    address mpContract,
+    address tallyOwner,
+    Mode pollMode
+  ) payable TallyBase(verifierContract, vkRegistryContract, pollContract, mpContract, tallyOwner, pollMode) {}
+
+  /// @notice A modifier that causes the function to revert if the cooldown period is not over
+  modifier afterCooldown() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration + cooldown) {
+      revert CooldownPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is over
+  modifier beforeVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed > duration) {
+      revert VotingPeriodOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is not over
+  modifier afterVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration) {
+      revert VotingPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the strategy is not initialized
+  modifier isInitialized() {
+    if (!initialized) {
+      revert NotInitialized();
+    }
+
+    _;
+  }
+
+  /// @notice Initialize tally with strategy params
+  /// @param params The strategy initialization params
+  function init(IPayoutStrategy.StrategyInit calldata params) public onlyOwner {
+    if (initialized) {
+      revert AlreadyInitialized();
+    }
+
+    if (params.poll != address(poll)) {
+      revert InvalidPoll();
+    }
+
+    initialized = true;
+    cooldown = params.cooldownTime;
+    registry = IPoll(params.poll).getRegistry();
+    token = IERC20(params.payoutToken);
+    maxContributionAmount = params.maxContribution;
+    voiceCreditFactor = params.maxContribution / MAX_VOICE_CREDITS;
+  }
+
+  /// @notice Pause contract calls (deposit, claim, withdraw)
+  function pause() public onlyOwner {
+    _pause();
+  }
+
+  /// @notice Unpause contract calls (deposit, claim, withdraw)
+  function unpause() public onlyOwner {
+    _unpause();
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function deposit(uint256 amount) public isInitialized whenNotPaused beforeVotingDeadline {
```

Comment:
this should probably be allowed until the tallying is completed (isTallied)

Reply:
I thought the logic here is the same as for message publishing. Is there a case when deposit is needed after poll is over?

Reply:
isn't deposit a function for the admin to add money to the matching pool?

Reply:
I think everyone can add money if they want :) In case if we need to restrict the access, I can add onlyOwner modifier.

Reply:
oh no it's fine open, what I meant is that we shouldn't limit this until the voting deadline ends, but allow until the ballots are tallied

Reply:
or even better if we prevent claim until all results are added, then we should be also allowing to add money until all results are added. This would allow for the operator to not deposit the matching funds until they are sure that the there are no errors in tallying (assuming there is an unknown circuit bug)

Reply:
So deposit function will be something like this?
```
function deposit(uint256 amount) public isInitialized whenNotPaused beforeTallying
```

Reply:
I would say let's do until before all results are posted. so we can allow them to upload after tallying all votes (accounting for something to break in the future in the circuits)